In [212]:
import pandas as pd
import numpy as np

In [213]:
file = ".//Data//Small_Data//offline_sales.json"

In [220]:
## Create DF with offline data
df = pd.read_json(file, lines=True, dtype={'customer_id': str})
## Read customer_id as string to deal with big numbers

# Exploratory Data Analysis

In [221]:
## Checking DF
df.head(10)

date state  store_id          sale_id   off_product_id  quantity  \
0 2018-08-01    RN   3162633  323932636465316  643839313230323         1   
1 2018-08-01    RN   3162633  326463633336323  613665646663346         1   
2 2018-08-01    RN   3162633  333331346330323  343739666265353         1   
3 2018-08-01    RN   3162633  393461383633336  373034633762613         1   
4 2018-08-01    RN   3162633  616531626536373  343634373062363         1   
5 2018-08-01    RN   3162633  626462353638383  616339353035393         1   
6 2018-08-01    RN   3162633  653363623539323  346165376638356         1   
7 2018-08-01    RN   3162633  666639323036376  643839313363376         1   
8 2018-08-01    RN   3162633  636238633865636  333835396465353         1   
9 2018-08-01    RN   3162633  626239383061353  343332613639373         1   

   price           customer_id  
0  279.0                  None  
1  229.0                  None  
2  179.0                  None  
3  269.0                  None  
4   29.0                  None  
5  459.0                  None  
6  369.0                  None  
7  149.0  30373934343338363136  
8  198.0  30393833353363303832  
9  229.0  31366262653431626130

In [226]:
## Replace None to NaN
df.customer_id.replace('None', np.nan, inplace=True)

In [227]:
## Total Entries
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29372 entries, 0 to 29371
Data columns (total 8 columns):
date              29372 non-null datetime64[ns]
state             29372 non-null object
store_id          29372 non-null int64
sale_id           29372 non-null int64
off_product_id    29372 non-null int64
quantity          29372 non-null int64
price             29372 non-null float64
customer_id       25844 non-null object
dtypes: datetime64[ns](1), float64(1), int64(4), object(2)
memory usage: 1.8+ MB


**29372 rows, 8 columns**

In [228]:
## Check dtypes
df.dtypes

date              datetime64[ns]
state                     object
store_id                   int64
sale_id                    int64
off_product_id             int64
quantity                   int64
price                    float64
customer_id               object
dtype: object

In [230]:
## Checking for NaN values
df.isnull().sum()

date                 0
state                0
store_id             0
sale_id              0
off_product_id       0
quantity             0
price                0
customer_id       3528
dtype: int64

**3528 NaN values on customer_id column**  
**Clean or manipulate this**  
**12% of costumer_id is Empty**

In [231]:
## Create DF with no null values (maybe use later)
no_null = df
no_null.dropna(axis=0, inplace=True)
no_null.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25844 entries, 7 to 29371
Data columns (total 8 columns):
date              25844 non-null datetime64[ns]
state             25844 non-null object
store_id          25844 non-null int64
sale_id           25844 non-null int64
off_product_id    25844 non-null int64
quantity          25844 non-null int64
price             25844 non-null float64
customer_id       25844 non-null object
dtypes: datetime64[ns](1), float64(1), int64(4), object(2)
memory usage: 1.8+ MB


In [232]:
## Drop old Index Column and Create a new one
no_null = no_null.reset_index().drop(columns=['index'])
no_null.head()

date state  store_id          sale_id   off_product_id  quantity  \
0 2018-08-01    RN   3162633  666639323036376  643839313363376         1   
1 2018-08-01    RN   3162633  636238633865636  333835396465353         1   
2 2018-08-01    RN   3162633  626239383061353  343332613639373         1   
3 2018-08-01    RN   3162633  376465626630613  306439363765386         1   
4 2018-08-01    RN   3162633  666564646366383  336139376666306         1   

   price           customer_id  
0  149.0  30373934343338363136  
1  198.0  30393833353363303832  
2  229.0  31366262653431626130  
3   58.0  32653962626632373464  
4   58.0  34653761646637306265